tools/src/main/java/org/apache/kafka/tools/TransactionalMessageCopier.java

```diff
@@ -0,0 +1,290 @@
+        final String outputTopic = parsedArgs.getString("outputTopic");
+
+        String consumerGroup = parsedArgs.getString("consumerGroup");
+        TopicPartition inputPartition = new TopicPartition(parsedArgs.getString("inputTopic"), parsedArgs.getInt("inputPartition"));
+
+        final KafkaProducer<String, String> producer = createProducer(parsedArgs);
+        final KafkaConsumer<String, String> consumer = createConsumer(parsedArgs, inputPartition);
+
+        consumer.assign(Arrays.asList(inputPartition));
+
+        long maxMessages = parsedArgs.getInt("maxMessages") == -1 ? Long.MAX_VALUE : parsedArgs.getInt("maxMessages");
+        maxMessages = Math.min(messagesRemaining(consumer, inputPartition), maxMessages);
+
+        producer.initTransactions();
+
+
+        final AtomicBoolean isShuttingDown = new AtomicBoolean(false);
+        final AtomicLong remainingMessages = new AtomicLong(maxMessages);
+        final AtomicLong numMessagesProcessed = new AtomicLong(0);
+        int exitCode = 0;
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                isShuttingDown.set(true);
+                // Flush any remaining messages
+                producer.close();
+                synchronized (consumer) {
+                    consumer.close();
+                }
+                System.out.println(shutDownString(numMessagesProcessed.get(), remainingMessages.get(), transactionalId));
+            }
+        });
+
+        try {
+            while (0 < remainingMessages.get()) {
+                if ((((double) numMessagesProcessed.get() / maxMessages) * 100) % 10 == 0) {
+                    // print status for every 10% we progress.
+                    System.out.println(statusAsJson(numMessagesProcessed.get(), remainingMessages.get(), transactionalId));
+                }
+                if (isShuttingDown.get())
+                    break;
+                int messagesInCurrentTransaction = 0;
+                long numMessagesForNextTransaction = Math.min(numMessagesPerTransaction, remainingMessages.get());
+                producer.beginTransaction();
+
+                while (messagesInCurrentTransaction < numMessagesForNextTransaction) {
+                    ConsumerRecords<String, String> records = consumer.poll(200L);
+                    for (ConsumerRecord<String, String> record : records) {
+                        producer.send(producerRecordFromConsumerRecord(outputTopic, record));
+                        messagesInCurrentTransaction++;
+                    }
+                }
+                producer.sendOffsetsToTransaction(consumerPositions(consumer), consumerGroup);
+                producer.commitTransaction();
+                remainingMessages.set(maxMessages - numMessagesProcessed.addAndGet(messagesInCurrentTransaction));
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
In `VerifiableProducer` and `VerifiableConsumer`, we pass uncaught exceptions to the argument parser `handleError`. It would be nice to be consistent.